### ISSUE TITLE
Support the image map processing for <img ismap/> inside an <a/>

### ISSUE THREAD

**Issue by shinglyu** [servo/servo#4873]
(This is a followup for #4236 )
Implement the __step 3__ for HTMLAnchorElement's activation behaviour spec: https://html.spec.whatwg.org/multipage/semantics.html#the-a-element:activation-behaviour

**Comment by shinglyu** [comment]
I'll take this one.

**Comment by shinglyu** [comment]
status update: Most of the things are ready, I can now identify if the clicked target is a img element, if it has a `ismap` property, and append the pointer coordinates to the url. The only piece missing is to get the `<img>`'s position and calculate the relative coordinate.

**Comment by shinglyu** [comment]
Hmm... do you think I can reuse the `reflow` code in `script/dom/window.rs`? Or do I have to write the whole thing for my purpose?

**Comment by jdm** [comment]
You shouldn't need to modify the `reflow` method; you can just create a new layout query and use it like `content_boxy_query` is used.

**Comment by shinglyu** [comment]
This is what I've done so far: https://github.com/shinglyu/servo/blob/ismap/components/script/dom/htmlanchorelement.rs#L140

Am I on the right direction? I wonder what this `content_box` is? The `min_x` and `min_y` is always 0 in my [simple test case](https://github.com/shinglyu/servo/blob/ismap/tests/html/test_img_ismap.html).

**Comment by jdm** [comment]
content_box_query is a query that just figures out the width and height of a element on a page. I meant that it should be a model for another query that figures out the x and y coordinates for an element.

**Comment by jdm** [comment]
In particular, the call to union rects in http://mxr.mozilla.org/servo/source/components/layout/layout_task.rs#1146 would cause the origin to never shift.

**Comment by shinglyu** [comment]
Cool, I'll start writing the new layout query

**Comment by jdm** [comment]
Actually, you might be able to make the content box query do this work by making it store an Option<Rect> and initialize it to the first one, rather than unioning with Rect::Zero. That would make the origin be set correctly.

**Comment by shinglyu** [comment]
1. What does the "first one" mean?
2. I tried my best, but still can't understand this "union with Rect::Zero" thing. What is the purpose of this `UnioningFragmentBorderBoxIterator`?  Is the fragment border box related to [this](http://dev.w3.org/csswg/css-break-3/)?

Thanks!

**Comment by jdm** [comment]
There are two parts to the `FragmentBorderBoxIterator` type - one is the method that is called for each matching fragment, and the other is the method that is called to decide if a fragment matches. By the "first one", I mean the rectangle that is retrieved the first time a matching fragment is found. `UnioningFragmentBorderBoxIterator` is used to find the union of all of the rectangles of the fragments that make up a DOM element. Right now it starts with a zero-sized rectangle, and the union operation makes that never move from the origin. If it started with `None` instead and assigned `Some(rect)` for the first rectangle and took the union for subsequent rectangles, I think that would do what you want.

**Comment by shinglyu** [comment]
OK. So let me rephrase it to check my understanding.

The purpose for the `context_box_query` is to get the width and height of an element. To do so, the layout task need to create a `Unioning FragmentBorderBoxIterator` with the requested node as parameter. Then the `sequential::iterate_through_flow_tree_fragment_border_boxes` will be called, which will iterate through the fragments and calculate the union of all the fragements that makes up the requested element, thus figuring out its size. Is that correct? 

So what I need to do is to change the starting rect of the `UnioningFragmentBorderBoxIterator` from a zero-sized rect to `None`, then replace it with the first matching rectangle. 

Questions: 
1. Does the "border box" in this context mean the bonding box of an element?
2. What does "match" mean?

Thanks!

**Comment by jdm** [comment]
Yes, the border box is the bounding box as far as I know. Match in this case means the rectangle that is passed to the `process` method of UnioningFragmentBorderBoxIterator, since `should_process` ensures that only rectangles that make up the fragment are processed.

**Comment by shinglyu** [comment]
(Sorry, the last week was a busy week)
I tried something like this: https://github.com/shinglyu/servo/blob/ismap/components/layout/layout_task.rs#L1079

But the `min_x` and `min_y` I get are still `0,0`. Also, the `debug!` in line 1177 does not print, not sure if I was modifying the wrong place.

**Comment by jdm** [comment]
I would expect the debug line to print if you use RUST_LOG=layout::layout_task, so that's quite strange. Those were the changes I was expecting :/

**Comment by shinglyu** [comment]
I did use `RUST_LOG=layout::layout_task`, even if I loosen up to use `RUST_LOG=layout`, I can't see the log.

**Comment by jdm** [comment]
Aha - I cloned your branch and tried it out, and #5458 is preventing any real value from being returned right now.

**Comment by shinglyu** [comment]
Thanks! I can never find that out myself. Let's see if I can fix #5458 then.

**Comment by jdm** [comment]
I submitted #5459 to fix it already; sorry.

**Comment by shinglyu** [comment]
Cool! Thanks!
